{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Derived from https://scanpy.readthedocs.io/en/stable/tutorials.html\n",
    "\n",
    "# Count matrices were generated using CellRanger Count (see 10X Genomics website)\n",
    "\n",
    "# import packages \n",
    "import scanpy.external as sce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "import seaborn as sb\n",
    "from gprofiler import gprofiler\n",
    "from seaborn import despine\n",
    "from seaborn import axes_style\n",
    "from matplotlib.pyplot import suptitle\n",
    "import magic # imputation tool; van Dijk et al 2018 # "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "\n",
    "# package versions used for this analysis\n",
    "# scanpy==1.4.4 anndata==0.6.22.post1 umap==0.3.9 numpy==1.17.2 scipy==1.3.0 pandas==0.23.4 scikit-learn==0.20.2 statsmodels==0.10.1 python-igraph==0.7.1 louvain==0.6.1\n",
    "\n",
    "#load and merge files\n",
    "\n",
    "filenames = [\n",
    "'AM1_filtered_feature_bc_matrix.h5',\n",
    "'AM2_filtered_feature_bc_matrix.h5',\n",
    "'AM4_filtered_feature_bc_matrix.h5',    \n",
    "'AM5_filtered_feature_bc_matrix.h5',\n",
    "'AM7_filtered_feature_bc_matrix.h5',  \n",
    "'AM9_filtered_feature_bc_matrix.h5']\n",
    "adatas = [sc.read_10x_h5(filename) for filename in filenames]\n",
    "\n",
    "adata = adatas[0].concatenate(adatas[1:],batch_categories=[\"AM1.ep_YFP_neg\",\"AM2.ep_YFP_pos\",\"AM4.ep_YFP_neg\",\"AM5.ep_YFP_pos\",\"AM7.ep_control\",\"AM9.ep_control\"])\n",
    "\n",
    "# make sure gene names are unique\n",
    "adata.var_names_make_unique()  \n",
    "\n",
    "# compute %mito and remove cells with >10%\n",
    "mito_genes = adata.var_names.str.startswith('mt-')\n",
    "\n",
    "# for each cell compute fraction of counts in mito genes vs. all genes\n",
    "# the `.A1` is only necessary as X is sparse (to transform to a dense array after summing)\n",
    "adata.obs['percent_mito'] = np.sum(adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
    "\n",
    "# add the total counts per cell as observations-annotation to adata\n",
    "adata.obs['n_counts'] = adata.X.sum(axis=1).A1\n",
    "\n",
    "# plot mitochondrial content (y-axis) against read count before and after filtering cells with > 10% mitochondrial gene expression\n",
    "# Cells wth >10% mito expression also has low read count indicating bad/dead cells rather than cells requiring more energy\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "with axes_style({'axes.grid': False}):\n",
    " sc.pl.scatter(adata, y='percent_mito', x='n_counts', size=5)\n",
    "\n",
    "adata = adata[adata.obs['percent_mito'] < 0.1, :]\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "with axes_style({'axes.grid': False}):\n",
    " sc.pl.scatter(adata, y='percent_mito', x='n_counts', size=5)\n",
    "\n",
    "# Normalize the data, save raw data, then use data diffusion tool (van Dijk et al 2018)\n",
    "sc.pp.normalize_per_cell(adata)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# No consensus about gene scaling so I did not scale the data. Read https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6582955/\n",
    "# Not scaling argueable retains biological information\n",
    "# I did not regress cell cycle either. A population of proliferative or non-proliferative cells in this context would be high interesting in my opinion, and is not information\n",
    "# I want to remove!\n",
    "\n",
    "# save current data as adata.raw\n",
    "adata.raw = adata\n",
    "\n",
    "# data diffusion tool is in scanpy.external \n",
    "sce.pp.magic(adata, name_list='all_genes', k=5, t=15, n_pca=20) # default settings\n",
    "\n",
    "#Intial filter\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "\n",
    "# create nearest neighbors graph and run louvain community detection algorithm\n",
    "sc.pp.neighbors(adata, n_neighbors=5, n_pcs=20)\n",
    "sc.tl.louvain(adata, resolution=0.05, key_added='louvain_r0.05')\n",
    "sc.tl.louvain(adata, resolution=0.1, key_added='louvain_r0.1')\n",
    "sc.tl.louvain(adata, resolution=0.2, key_added='louvain_r0.2')\n",
    "sc.tl.louvain(adata, resolution=0.3, key_added='louvain_r0.3')\n",
    "sc.tl.louvain(adata, resolution=0.4, key_added='louvain_r0.4')\n",
    "sc.tl.louvain(adata, resolution=0.5, key_added='louvain_r0.5')\n",
    "\n",
    "# Visualize different Louvain resolutions using UMAP\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "sc.pl.umap(adata, color=['batch','louvain_r0.05','louvain_r0.1','louvain_r0.2','louvain_r0.3','louvain_r0.4','louvain_r0.5'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a PAGA-initialized UMAP embedding (Wolf et al. 2018 Genome Biology). Louvain_r0.05 was chosen\n",
    "\n",
    "sc.tl.paga(adata, groups='louvain_r0.05')\n",
    "sc.pl.paga(adata, plot=True, color=['batch'])\n",
    "\n",
    "# sc.tl.umap(adata, init_pos='paga') is not working (August 7, 2019) https://github.com/theislab/scanpy/issues/769 = work around for now..\n",
    "\n",
    "sc.tl.umap(adata, init_pos=sci.tl._utils.get_init_pos_from_paga(adata))\n",
    "sc.pl.umap(adata, frameon=True, color=['louvain_r0.05'], legend_loc='right margin', size=40) # louvain clusters, resolution = 0.05\n",
    "sc.pl.umap(adata, frameon=True, color=['louvain_r0.05'], legend_loc='right margin', size=40) # batch\n",
    "sc.pl.umap(adata, frameon=True, color=['Lyz2','Scgb1a1','Scgb3a2','Foxj1','Ly6a'],  color_map='RdYlBu_r', size=40, ncols=2, use_raw=False, vmin=0) # distal lung epithelial markers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove louvain clusters with less than 300 cells\n",
    "adata.obs['louvain_r0.05'].value_counts()\n",
    "adata_subset = adata[adata.obs['louvain_r0.05'].isin(['0','1','2','3','4','5'])]\n",
    "subset_results_file = './write/adata_subset.h5ad'  # the file that will store the analysis results\n",
    "adata_subset.write(subset_results_file) # save subset data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the contribution of each library ID to the remaining louvain clusters\n",
    "rcParams['figure.figsize'] = 8,5\n",
    "with axes_style({'axes.grid': False}):\n",
    " ax = sb.countplot(x=\"louvain_r0.05\", hue=\"batch\", data=adata_subset.obs, linewidth=0.5, edgecolor='black')\n",
    " despine(right=True)\n",
    "plt.xlabel(\"Louvain communities\")\n",
    "plt.ylabel(\"Cells\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Differential expression (DE) and Enrichr analysis\n",
    "\n",
    "# Identify DE gene using the in-built ScanPy function\n",
    "\n",
    "# Filtered gene list for TF/TFC identification. I wanted to be stringent\n",
    "sc.tl.rank_genes_groups(adata_subset, 'louvain_r0.05', method='wilcoxon', n_genes=1000, use_raw=True)\n",
    "\n",
    "# Removed the outgroup filter, but kept the log fold change and ingroup filter\n",
    "sc.tl.filter_rank_genes_groups(adata_subset, \n",
    "   min_fold_change = 1, # minimum log fold change   \n",
    "   min_in_group_fraction = 0.5,  # 50% of cells in the cluster must express the gene\n",
    "   max_out_group_fraction = 0.5, # No more than 50% of cells outside a cluster can express this genes\n",
    "   key_added='rank_genes_groups_filtered') # filtered group key\n",
    "\n",
    "result = adata_subset.uns['rank_genes_groups_filtered']\n",
    "groups = result['names'].dtype.names\n",
    "\n",
    "# export filtered DE results to excel \n",
    "df1=pd.DataFrame({group + '_' + key[:1]: result[key][group] for group in groups for key in ['names', 'pvals_adj','logfoldchanges']})\n",
    "df1.to_excel(\"adata_subset_DE_top1000_filtered.xlsx\", sheet_name='Top 1000 filtered')\n",
    "\n",
    "# Visualize the data using a heatmap\n",
    "heatmap_cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", ['lightblue','lightyellow','lightcoral']) # color map\n",
    "\n",
    "sc.pl.rank_genes_groups_heatmap(adata_subset, n_genes=25,cmap=heatmap_cmap,swap_axes=True, use_raw=False, key='rank_genes_groups_filtered', vmin=0, vmax=1.5, show_gene_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify all DE TF's \n",
    "\n",
    "# import list of mouse TFs from TFDB\n",
    "# http://bioinfo.life.hust.edu.cn/AnimalTFDB/#!/species\n",
    "df_tf = pd.read_csv('Mus_musculus_TF.csv', header=0)\n",
    "tf_genes = set(df_tf['Symbol'].tolist())\n",
    "\n",
    "# Find TF/TFCs in the filtered DE genes list\n",
    "diff_padj_filtered = pd.read_excel('adata_subset_DE_top1000_filtered.xlsx', header = 0) # previously generated using in built ScanPy function\n",
    "names = diff_padj_filtered.columns\n",
    "clust_dict = {} #key,value\n",
    "\n",
    "for col in diff_padj_filtered.columns:\n",
    "    \n",
    "    # get diff gene list\n",
    "    cluster_genes = diff_padj_filtered[col].tolist()\n",
    "\n",
    "    # Create results array\n",
    "    results_tfs = []\n",
    "\n",
    "    # iterate through and pull out TF/TFCs\n",
    "    \n",
    "    for i in tf_genes:\n",
    "     if i in cluster_genes:\n",
    "      results_tfs.append(i)\n",
    "    \n",
    "    #store in dict (receptors,ligands,tfs)\n",
    "    the_key = col\n",
    "    the_value = {}\n",
    "    the_value[\"tfs\"] = results_tfs\n",
    "    clust_dict[the_key] = the_value\n",
    "\n",
    "# Create a list of identified TFs\n",
    "merged_tf = []\n",
    "\n",
    "for i in clust_dict:\n",
    "    sub_dict = clust_dict[i] #access each cluster in dict\n",
    "    \n",
    "    # create merge\n",
    "    merged_tf += sub_dict[\"tfs\"]\n",
    "\n",
    "# Visualize data as a matrix plot\n",
    "\n",
    "cmap2 = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"white\",'aliceblue',\"royalblue\"])\n",
    "\n",
    "sc.pl.matrixplot(adata_subset, var_names=de_tfs, cmap=cmap2, groupby='louvain_r0.05', use_raw=False, swap_axes=True, vmin=0, vmax=0.6, dendrogram=False)\n",
    "\n",
    "# You can also check TF/TFCs per cluster. For example:\n",
    "clust_dict['0_n'][\"tfs\"] \n",
    "\n",
    "# Unfiltered gene list for TF/TFC identification. I wanted to be less stringent here\n",
    "sc.tl.rank_genes_groups(adata_subset, 'louvain_r0.05', method='wilcoxon',n_genes=1000, use_raw=True)\n",
    "\n",
    "result = adata_subset.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "\n",
    "# export DE results to excel \n",
    "mouse_DE_df=pd.DataFrame({group + '_' + key[:1]: result[key][group] for group in groups for key in ['names', 'pvals_adj','logfoldchanges']})\n",
    "mouse_DE_df.to_excel(\"adata_subset_DE_top1000_unfiltered.xlsx\", sheet_name='Top 1000 unfiltered')\n",
    "\n",
    "# Look for enriched Gene Ontology Biological Process 2018 pathways\n",
    "\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina' # mac\n",
    "import gseapy as gp\n",
    "from gseapy.plot import barplot, dotplot # only needed if you want bar and dot plots\n",
    "\n",
    "#view available reference libraries\n",
    "names = gp.get_library_name() # a list of available libraries will appear\n",
    "\n",
    "# turn columns into lists (filtered data)\n",
    "C0 = mouse_DE_df['0_n']\n",
    "C1 = mouse_DE_df['1_n']\n",
    "C2 = mouse_DE_df['2_n']\n",
    "C3 = mouse_DE_df['3_n']\n",
    "C4 = mouse_DE_df['4_n']\n",
    "C5 = mouse_DE_df['5_n']\n",
    "\n",
    "# drop NaN: NaNs cause enricher to break\n",
    "C0=C0.dropna()\n",
    "C1=C1.dropna()\n",
    "C2=C2.dropna()\n",
    "C3=C3.dropna()\n",
    "C4=C4.dropna()\n",
    "C5=C5.dropna()\n",
    "\n",
    "# GO Analysis\n",
    "\n",
    "Cluster0_GOBio = gp.enrichr(gene_list = C0,\n",
    "description='Cluster0_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster0_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "Cluster1_GOBio = gp.enrichr(gene_list = C1,\n",
    "description='Cluster1_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster1_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "Cluster2_GOBio = gp.enrichr(gene_list = C2,\n",
    "description='Cluster2_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster2_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "Cluster3_GOBio = gp.enrichr(gene_list = C3,\n",
    "description='Cluster3_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster3_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "Cluster4_GOBio = gp.enrichr(gene_list = C4,\n",
    "description='Cluster4_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster4_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "Cluster5_GOBio = gp.enrichr(gene_list = C5,\n",
    "description='Cluster5_GOBio',\n",
    "gene_sets=['GO_Biological_Process_2018'],\n",
    "outdir='Enricher_analysis/Cluster5_GOBio',\n",
    "cutoff=0.05\n",
    ")\n",
    "\n",
    "# find unique GO terms in clusters 0, \n",
    "\n",
    "# filter for statistically significant terms\n",
    "\n",
    "c0_sig = Cluster0_GOBio.res2d.loc[(Cluster0_GOBio.res2d['Adjusted P-value'] < 0.05)] \n",
    "c1_sig = Cluster1_GOBio.res2d.loc[(Cluster1_GOBio.res2d['Adjusted P-value'] < 0.05)] \n",
    "\n",
    "c0_list = c0_sig['Term'].tolist() # wt\n",
    "c1_list = c1_sig['Term'].tolist() # cancer\n",
    "\n",
    "# look for unique and common terms in C1 and C0\n",
    "\n",
    "at2_c0_unique_terms = []\n",
    "at2_c1_unique_terms = []\n",
    "at2_common_terms = []\n",
    "\n",
    "for i in c0_list:\n",
    "    if i not in c1_list:\n",
    "        at2_c0_unique_terms.append(i)\n",
    "\n",
    "for i in c1_list:\n",
    "    if i not in c0_list:\n",
    "        at2_c1_unique_terms.append(i)\n",
    "       \n",
    "for i in c1_list:\n",
    "    if i in c0_list:\n",
    "        at2_common_terms.append(i)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export  unique terms to excel\n",
    "pd.DataFrame(c0_unique_terms).to_excel(\"Unique cluster 0 terms.xlsx\", sheet_name='Sheet_name_1')\n",
    "pd.DataFrame(c1_unique_terms).to_excel(\"Unique cluster 1 terms.xlsx\", sheet_name='Sheet_name_1')\n",
    "pd.DataFrame(c3_unique_terms).to_excel(\"Unique cluster 3 terms.xlsx\", sheet_name='Sheet_name_1')\n",
    "pd.DataFrame(common_terms).to_excel(\"common terms.xlsx\", sheet_name='Sheet_name_1')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot overlap using a venn diagram and check it matches the length of the lists above\n",
    "\n",
    "from matplotlib_venn import venn2\n",
    "c=venn2([set(c0_list), set(c1_list)], set_labels = ('C0', 'C1'))\n",
    "plt.show() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate gene signatures. Principle is true for all gene lists provided in supplementary\n",
    "\n",
    "# Calculate Kras score using Bild et al. Nature Kras signature. Also used in Barbie et al.\n",
    "\n",
    "bild_kras = pd.read_excel('Supplementary Table 2.xlsx', sheet_name='Bild et al. 2006 (Kras genes)', header=0) # load lists \n",
    "\n",
    "bild_kras_list = bild_kras['GeneSymbol'].tolist() # Make list of gene names\n",
    "\n",
    "# Gene names are capitals. Lower and capitalize to be compatible with dataset gene names\n",
    "\n",
    "bild_kras_list = [x.lower() for x in bild_kras_list] # lower\n",
    "bild_kras_list = [x.capitalize() for x in bild_kras_list] # capitalize\n",
    "\n",
    "bild_kras_list_final = [x for x in bild_kras_list if x in adata.var_names] # remove genes not in adata.var\n",
    "\n",
    "# Calculating z-scores for single cells\n",
    "# Derived from https://github.com/theislab/scanpy/issues/181\n",
    "\n",
    "# Create a cell barcodes column\n",
    "adata_subset.obs['index1'] = adata_subset.obs.index\n",
    "\n",
    "# create the marker dict \n",
    "marker_dict = dict()\n",
    "marker_dict['Kras z-score'] = bild_kras_list_final\n",
    "\n",
    "# create the function\n",
    "def evaluate_partition(anndata, marker_dict, gene_symbol_key=None, partition_key=None):\n",
    "    # Inputs:\n",
    "    #    anndata         - An AnnData object containing the data set and a partition\n",
    "    #    marker_dict     - A dictionary with cell-type markers. The markers should be stores as anndata.var_names or \n",
    "    #                      an anndata.var field with the key given by the gene_symbol_key input\n",
    "    #    gene_symbol_key - The key for the anndata.var field with gene IDs or names that correspond to the marker \n",
    "    #                      genes\n",
    "    #    partition_key   - The key for the anndata.obs field where the cluster IDs are stored. The default is\n",
    "    #                      'louvain_r1' \n",
    "\n",
    "    #Test inputs\n",
    "    if partition_key not in anndata.obs.columns.values:\n",
    "        print('KeyError: The partition key was not found in the passed AnnData object.')\n",
    "        print('   Have you done the clustering? If so, please tell pass the cluster IDs with the AnnData object!')\n",
    "        raise\n",
    "\n",
    "    if (gene_symbol_key != None) and (gene_symbol_key not in anndata.var.columns.values):\n",
    "        print('KeyError: The provided gene symbol key was not found in the passed AnnData object.')\n",
    "        print('   Check that your cell type markers are given in a format that your anndata object knows!')\n",
    "        raise\n",
    "        \n",
    "    if gene_symbol_key:\n",
    "        gene_ids = anndata.var[gene_symbol_key]\n",
    "    else:\n",
    "        gene_ids = anndata.var_names\n",
    "        \n",
    "    # Create a column based on index. This allows z-score calculation on single cells rather than clusters\n",
    "    clusters = np.unique(anndata.obs[partition_key])\n",
    "    n_clust = len(clusters)\n",
    "    n_groups = len(marker_dict)\n",
    "    \n",
    "    marker_res = np.zeros((n_groups, n_clust))\n",
    "    z_scores = sc.pp.scale(anndata, copy=True) # try changing this to anndata_raw as a separate function\n",
    "\n",
    "    i = 0\n",
    "    for group in marker_dict:\n",
    "        # Find the corresponding columns and get their mean expression in the cluster\n",
    "        j = 0\n",
    "        for clust in clusters:\n",
    "            cluster_cells = np.in1d(z_scores.obs[partition_key], clust)\n",
    "            marker_genes = np.in1d(gene_ids, marker_dict[group])\n",
    "            marker_res[i,j] = z_scores.X[np.ix_(cluster_cells,marker_genes)].mean()\n",
    "            j += 1\n",
    "        i+=1\n",
    "\n",
    "    variances = np.nanvar(marker_res, axis=0)\n",
    "    if np.all(np.isnan(variances)):\n",
    "        print(\"No variances could be computed, check if your cell markers are in the data set.\")\n",
    "        print(\"Maybe the cell marker IDs do not correspond to your gene_symbol_key input or the var_names\")\n",
    "        raise\n",
    "\n",
    "    marker_res_df = pd.DataFrame(marker_res, columns=clusters, index=marker_dict.keys())\n",
    "    \n",
    "    return marker_res_df\n",
    "\n",
    "   # Return the median of all the variances over the clusters\n",
    "    #marker_matches = ([np.median(variances), marker_res_df])\n",
    "    \n",
    "   # return marker_matches\n",
    "\n",
    "# Calculate the z-score\n",
    "df = evaluate_partition(adata_subset, marker_dict, gene_symbol_key=None, partition_key = 'index1')\n",
    "\n",
    "# Transpose the dataframe\n",
    "df_transposed = df.transpose()\n",
    "\n",
    "# add score to adata.obs\n",
    "adata_subset.obs['Kras z-score'] = df_transposed['Kras z-score']\n",
    "\n",
    "# save the data\n",
    "adata_subset.write(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### You can download the annotated adata_subset file and begin data analysis straight away! ####\n",
    "\n",
    "# load adata_subset file shown in the figures\n",
    "results_file = 'adata_subset.h5ad'  \n",
    "adata_subset = sc.read(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hope the data and code were useful!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
